sidechain/consensus/aura/src/peer_block_sync.rs

```diff
@@ -0,0 +1,193 @@
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::block_production_suspension::{IsBlockProductionSuspended, SuspendBlockProduction};
+use core::marker::PhantomData;
+use its_consensus_common::{BlockImport, Error, Result};
+use its_primitives::{traits::SignedBlock as SignedSidechainBlockTrait, types::BlockHash};
+use log::*;
+use sp_runtime::traits::{Block as ParentchainBlockTrait, Header as ParentchainHeaderTrait};
+use std::{sync::Arc, vec::Vec};
+
+/// Trait for syncing sidechain blocks from a peer validateer.
+///
+/// This entails importing blocks and detecting if we're out of date with our blocks, in which
+/// case we fetch the missing blocks from a peer.
+pub trait SyncBlockFromPeer<ParentchainHeader, SignedSidechainBlock>
+where
+	ParentchainHeader: ParentchainHeaderTrait,
+	SignedSidechainBlock: SignedSidechainBlockTrait,
+{
+	fn attempt_block_sync(
+		&self,
+		sidechain_block: SignedSidechainBlock,
+		last_imported_parentchain_header: &ParentchainHeader,
+	) -> Result<()>;
+}
+
+/// Sidechain peer block sync implementation.
+pub struct PeerBlockSync<
+	ParentchainBlock,
+	SignedSidechainBlock,
+	BlockImporter,
+	BlockProductionSuspender,
+> {
+	importer: Arc<BlockImporter>,
+	block_production_suspender: Arc<BlockProductionSuspender>,
+	_phantom: PhantomData<(ParentchainBlock, SignedSidechainBlock)>,
+}
+
+impl<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+	PeerBlockSync<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+where
+	ParentchainBlock: ParentchainBlockTrait,
+	SignedSidechainBlock: SignedSidechainBlockTrait,
+	BlockImporter: BlockImport<ParentchainBlock, SignedSidechainBlock>,
+	BlockProductionSuspender: SuspendBlockProduction + IsBlockProductionSuspended,
+{
+	pub fn new(
+		importer: Arc<BlockImporter>,
+		block_production_suspender: Arc<BlockProductionSuspender>,
+	) -> Self {
+		PeerBlockSync { importer, block_production_suspender, _phantom: Default::default() }
+	}
+
+	fn sync_blocks_from_peer(
+		&self,
+		_last_known_block_hash: BlockHash,
+		last_imported_parentchain_header: &ParentchainBlock::Header,
+	) -> Result<()> {
+		// TODO fetch blocks from peer
+
+		let blocks_to_import: Vec<SignedSidechainBlock> = Vec::new();
+		for block_to_import in blocks_to_import {
+			if let Err(e) =
+				self.importer.import_block(block_to_import, last_imported_parentchain_header)
+			{
+				// If we encounter an error here, we have to abort the sync process, no way to recover.
+				error!("Error while importing a block during peer sync process, cannot recover, have to abort");
+				return Err(e)
+			}
+		}
+
+		Ok(())
+	}
```

Comment:
In this function we fetch the blocks from a peer and import them. Still TODO, because we don't have the whole o-call stack yet that does the fetching.